README.md
# Project Wycheproof

https://github.com/google/wycheproof

_Project Wycheproof is named after
[Mount Wycheproof](https://en.wikipedia.org/wiki/Mount_Wycheproof), the smallest
mountain in the world. The main motivation for the project is to have a goal
that is achievable. The smaller the mountain the more likely it is to be able to
climb it._

[TOC]

## Introduction

Project Wycheproof tests crypto libraries against known attacks. It is developed
and maintained by members of Google Security Team, but it is not an official
Google product.

At Google, we rely on many third party cryptographic software libraries.
Unfortunately, in cryptography, subtle mistakes can have catastrophic
consequences, and we found that libraries fall into such implementation pitfalls
much too often and for much too long. Good implementation guidelines, however,
are hard to come by: understanding how to implement cryptography securely
requires digesting decades' worth of academic literature. We recognize that
software engineers fix and prevent bugs with unit testing, and we found that
cryptographic loopholes can be resolved by the same means.

These observations have prompted us to develop Project Wycheproof, a collection
of unit tests that detect known weaknesses or check for expected behaviors of
some cryptographic algorithm. Project Wycheproof provides tests for most
cryptographic algorithms, including RSA, elliptic curve crypto and authenticated
encryption. Our cryptographers have systematically surveyed the literature and
implemented most known attacks. We have over 80 test cases which have uncovered
more than [40 bugs](doc/bugs.md). For example, we found that we could recover
the private key of widely-used DSA and ECDHC implementations.

While we are committed to develop as many attacks as possible, Project
Wycheproof is by no means complete. Passing the tests does not imply that the
library is secure, it just means that it is not vulnerable to the attacks that
Project Wycheproof tests for. Cryptographers are also constantly discovering new
attacks. Nevertheless, with Project Wycheproof developers and users now can
check their libraries against a large number of known attacks, without having to
spend years reading academic papers or become cryptographers themselves.

For more information on the goals and strategies of Project Wycheproof, please
check out our [doc](doc/).

### Coverage

Project Wycheproof has tests for the most popular crypto algorithms, including

- AES-EAX
- AES-GCM
- [DH](doc/dh.md)
- DHIES
- [DSA](doc/dsa.md)
- [ECDH](doc/ecdh.md)
- ECDSA
- ECIES
- [RSA](doc/rsa.md)

The tests detect whether a library is vulnerable to many attacks, including

- Invalid curve attacks
- Biased nonces in digital signature schemes
- Of course, all Bleichenbacher’s attacks
- And many more -- we have over 80 test cases

Our first set of tests are written in Java, because Java has a common
cryptographic interface. This allowed us to test multiple providers with a
single test suite. While this interface is somewhat low level, and should not be
used directly, we still apply a "defense in depth" argument and expect that the
implementations are as robust as possible. For example, we consider weak default
values to be a significant security flaw. We are converting as many tests into
sets of test vectors to simplify porting the tests to other languages. We
provide ready-to-use test runners for Java Cryptography Architecture providers
such as [Bouncy Castle](http://bouncycastle.org),
[Spongy Castle](https://rtyley.github.io/spongycastle/), the
[Amazon Corretto Crypto Provider](https://github.com/corretto/amazon-corretto-crypto-provider)
and the default providers in [OpenJDK](http://openjdk.java.net/).

### Usage

- Install [Bazel](https://bazel.build/).

- Install
  [Java Cryptography Extension (JCE) Unlimited Strength Jurisdiction Policy Files](http://stackoverflow.com/questions/6481627/java-security-illegal-key-size-or-default-parameters):
  this enables tests with large key sizes. Otherwise you'll see a lot of
  "illegal key size" exceptions.

- Check out the tests

```
git clone https://github.com/google/wycheproof.git
```

- To test latest stable version of Bouncy Castle:

```
bazel test BouncyCastleAllTests
```

- To test other versions, e.g., v1.52:

```
bazel test BouncyCastleAllTests_1_52
```

- To test all known versions (warning, will take a long time):

```
bazel test BouncyCastleAllTests_*
```

- To test a local jar, set the `WYCHEPROOF_BOUNCYCASTLE_JAR` environment
  variable:

```shell
$ WYCHEPROOF_BOUNCYCASTLE_JAR=/path/to/bouncycastle
$ bazel test BouncyCastleTestLocal
$ bazel test BouncyCastleAllTestsLocal
```

Note: bazel does not currently invalidate the build on environment changes. If
you change the `WYCHEPROOF_BOUNCYCASTLE_JAR` environment variable, run
`bazel clean` to force a rebuild:

```shell
$ WYCHEPROOF_BOUNCYCASTLE_JAR=/path/to/bouncycastle
$ bazel test BouncyCastleTestLocal
$ WYCHEPROOF_BOUNCYCASTLE_JAR=/path/to/other/jar
$ bazel clean
$ bazel test BouncyCastleTestLocal
```

- To test [Spongy Castle](https://rtyley.github.io/spongycastle/), replace
  BouncyCastle with SpongyCastle in your commands, for example

```
bazel test SpongyCastleAllTests
```

- To test the
  [Amazon Corretto Crypto Provider](https://github.com/corretto/amazon-corretto-crypto-provider),
  replace BouncyCastle with Accp in your commands, for example

```
bazel test AccpAllTests
```

- To test a local jar for the Amazon Corretto Crypto Provider, set the
  `WYCHEPROOF_ACCP_JAR` environment variable:

```shell
$ WYCHEPROOF_ACCP_JAR=/path/to/accp
$ bazel test AccpTestLocal
$ bazel test AccpAllTestsLocal
```

Note: bazel does not currently invalidate the build on environment changes. If
you change the `WYCHEPROOF_ACCP_JAR` environment variable, run `bazel clean` to
force a rebuild:

```shell
$ WYCHEPROOF_ACCP_JAR=/path/to/accp
$ bazel test AccpTestLocal
$ WYCHEPROOF_ACCP_JAR=/path/to/other/jar
$ bazel clean
$ bazel test AccpTestLocal
```

- To test your current installation of [OpenJDK](http://openjdk.java.net/):

```
bazel test OpenJDKAllTests
```

Note that OpenJDKAllTests expects that OpenJDK is your default JDK, so it might
refuse to run or its results might be incorrect if you are using some other JDK.
If you downloaded your JDK from Oracle or https://java.com, you're probably
using Oracle JDK, which should be compatible with OpenJDK, thus the tests should
run correctly.

Some tests take a very long time to finish. If you want to exclude them, use
BouncyCastleTest, SpongyCastleTest or OpenJDKTest -- these targets exclude all
slow tests (which are annotated with @SlowTest).

Most test targets are failing, and each failure might be a security issue. To
learn more about what a failed test means, you might want to check out
[our documentation](doc/bugs.md) or the comments on top of the corresponding
test function and test class.

### Hall of Bugs

Here are some of the notable vulnerabilities that are uncovered by Project
Wycheproof:

- OpenJDK's SHA1withDSA leaks private keys > 1024 bits

  - Test: testBiasSha1WithDSA in
    [DsaTest](https://github.com/google/wycheproof/blob/master/java/com/google/security/wycheproof/testcases/DsaTest.java).
  - This bug is the same as
    [CVE-2003-0971 - GnuPG generated ElGamal signatures that leaked the private key](https://web.nvd.nist.gov/view/vuln/detail?vulnId=CVE-2003-0971).

- Bouncy Castle's ECDHC leaks private keys
  - Test: testModifiedPublic and testWrongOrderEcdhc in
    [EcdhTest](https://github.com/google/wycheproof/blob/master/java/com/google/security/wycheproof/testcases/EcdhTest.java).

### Maintainers

Project Wycheproof is maintained by:

- Daniel Bleichenbacher
- Thai Duong
- Emilia Kasper
- Quan Nguyen

### Contact and mailing list

If you want to contribute, please read [CONTRIBUTING](CONTRIBUTING.md) and send
us pull requests. You can also report bugs or request new tests.

If you'd like to talk to our developers or get notified about major new tests,
you may want to subscribe to our
[mailing list](https://groups.google.com/forum/#!forum/wycheproof-users). To
join, simply send an empty mail to wycheproof-users+subscribe@googlegroups.com.


doc/changes.md
## Changes

### Version 0.6

- Added test vectors for RSA-OAEP decryption. The files are:
  `rsa_oaep_<keysize>_<hash>_<mgf>_test.json`.
- Added test vectors for ECDSA signatures with P1363 encoding. The files are:
  `ecdsa_<curve>_<hash>_p1363_test.json`
- Added test vectors for XDH with ASN encoded keys. The files are:
  `x25519_asn_test.json` and `x448_asn_test.json`
- Added test vectors for RSA-PKCS #1 signature generation. The current test
  vectors for RSA-PKCS #1 are meant for testing signature verification. These
  test vectors for signature generation include a private key. The files are:
  `rsa_sig_gen_misc_test.json`
- Removed a duplicate field "curve" in the test vectors for X25519. The "curve"
  field always describes the curve of the private key. All test vectors in the
  same test group (and in fact the same test file) use the same curve.
- Adding labels for all test vectors with status "acceptable". Test vectors have
  a status "acceptable" when it is unclear if a library should or should not
  accept them. This happens when the parameters are weak (e.g. using SHA-1 for
  signatures) or when the encoding is slightly non-standard.

### Version 0.7

Format changes\:

- All test vector files contain a new field `"schema"`, which points to a JSON
  schema definition in the directory `wycheproof/schema/`.
- More consistent type definition: E.g. the field `"type"` in a `testGroup` is
  now describing the type of the test the vectors are intended for
  [types.md](types.md). The type of the test defines formatting (e.g. whether
  signatures use P1363 encoding or ASN encoding) and the tested operation (e.g.
  whether the test vectors signatures are primarily are meant for verification
  or signature generation).
- RSA keys have been extended to use multi prime variants. The key material is
  now in a separate structure.
- The format of test vectors have been slightly unified: In particular,
  RSA-PKCS#1 v1.5 signatures no longer have a field "padding".
- Public and private keys in jwk format were added to other formats when the jwk
  format for these keys is defined.
- The size of the encoding of a BigInteger is now always a multiple of 2 to make
  the enoding closer to the encoding used in ASN.1. (It still uses twos
  complement, bigendian, hexadecimal endoding).

Additional test vectors\:

- Added vectors for edge cases that can occur in ECDH computations with
  projective or Jacobian coordinates. In particular this implement a detection
  for the attack in "Zero-Value Point Attacks on Elliptic Curve Cryptosystem" by
  T.Akishita and T Takagi, ISC 2003.
- Added more edge cases for Xdh.
- Added more test cases for ASN parsing (e.g. high number tags)
- Added test vectors for CVE-2017-18330 to CCM and EAX.
- Added more edge cases for poly1305.
- Added test vectors for HKDF.
- Added new test vectors for ED448.
- Added some test vectors using SHA-3.
- Added documentation of the tests intended by the test vector files:
  [files.md](files.md)
- Added test vectors using three prime RSA keys.
- Removed some duplicates in the test vectors.

### Some potential plans for version 0.8

- Adding a component for the analysis of key generation and signature
  generation. Currently, such tests only run against Java providers. The goal is
  to allow any library to be tested.
- Testing RSA-key generation. E.g. generate some RSA keys, then check if the
  keys have patterns that can be used by special case factoring algorithms.
- Extending tests for timing differences. So far there are only a limited number
  of tests against Java providers. The statistical analysis of the measurments
  are relatively simple and could probably be improved. One goal is to separate
  measuring the timings and doing the statistical analysis by defining a JSON
  structure for storing the measurements.
- Adding test vectors for long hashes and MACs. I.e. the inputs are strings and
  number of repetitions. This allows to add tests for primitives like
  HMAC('a' \* (2\*\*32+12345)).
- New algorithms: PMAC, X963KDF

### Some potential plans for later versions

- Adding code for capabilities of the providers. Currently this is written in
  unstructured form into the logs. Better would be a standalone binary that
  writes, JSON or html or text.
- alternative protocols: PKCS #11, AMD SEV, ...


javascript/webcryptoapi/testcases/EcdsaTest.js
/**
 * @license
 * Copyright 2017 Google Inc. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Tests for ECDSA implementations of Web Crypto API.
 */
goog.provide("wycheproof.webcryptoapi.ECDSA");
goog.require("e2e.ecc.Ecdsa");
goog.require("goog.testing.TestCase");
goog.require("goog.testing.asserts");
goog.require("goog.testing.jsunit");
goog.require("wycheproof.BigInteger");
goog.require("wycheproof.EcUtil");
goog.require("wycheproof.TestUtil");
goog.require("wycheproof.webcryptoapi.HashUtil");

var TestUtil = wycheproof.TestUtil;
var EcUtil = wycheproof.EcUtil;
var BigInteger = wycheproof.BigInteger;
var HashUtil = wycheproof.webcryptoapi.HashUtil;

// Test vector files
var ECDSA_VECTOR_FILE = "../../testvectors/ecdsa_webcrypto_test.json";

/** ECDSA wrapper */
var Ecdsa = function() {};

/**
 * Verifies the given signature using the given ECDSA public key.
 * @param {!CryptoKey} pk The ECDSA public key
 * @param {string} hashAlg The hash algorithm
 * @param {!ArrayBuffer} msg The message to be verified
 * @param {!ArrayBuffer} sig The signature to be verified
 *
 * @return {!Promise}
 */
Ecdsa.verify = function(pk, hashAlg, msg, sig) {
  return crypto.subtle.verify(
    { name: "ECDSA", hash: { name: hashAlg } },
    pk,
    sig,
    msg
  );
};

/**
 * Signs a message using the given ECDSA private key and hash algorithm.
 * @param {!CryptoKey} sk The ECDSA private key
 * @param {!ArrayBuffer} msg The message to be signed
 * @param {string} hashAlg The hash algorithm
 *
 * @return {!Promise}
 */
Ecdsa.sign = function(sk, msg, hashAlg) {
  return crypto.subtle.sign(
    {
      name: "ECDSA",
      hash: { name: hashAlg }
    },
    sk,
    msg
  );
};

/**
 * Imports a ECDSA public key.
 * @param {!JSONObject} keyData The key data in JWK format
 * @param {string} hashAlg The hash algorithm
 * @param {!Array<string>} usages The usages of the key
 *
 * @return {!Promise}
 */
Ecdsa.importPublicKey = function(keyData, hashAlg, usages) {
  return crypto.subtle.importKey(
    "jwk",
    keyData,
    { name: "ECDSA", namedCurve: keyData["crv"], hash: { name: hashAlg } },
    true,
    usages
  );
};

/**
 * Exports the given ECDSA key as a JSON object.
 * @param {!CryptoKey} key The ECDSA key to be exported
 *
 * @return {!Promise}
 */
Ecdsa.exportKey = function(key) {
  return crypto.subtle.exportKey("jwk", key);
};

/**
 * Generates an ECDSA key pair using the given hash algorithm and curve name.
 * @param {string} hashAlg The hash algorithm
 * @param {string} curveName The curve name
 *
 * @return {!Promise}
 */
Ecdsa.generateKey = function(hashAlg, curveName) {
  return crypto.subtle.generateKey(
    {
      name: "ECDSA",
      namedCurve: curveName
    },
    true,
    ["sign", "verify"]
  );
};

/**
 * Tests ECDSA signature verification. The test case's parameters are passed
 * to 'this' variable of the function.
 *
 * @return {!Promise}
 */
Ecdsa.testVerify = function() {
  tc = this;
  var promise = new Promise((resolve, reject) => {
    Ecdsa.importPublicKey(tc.keyData, tc.hashAlg, ["verify"])
      .then(function(pk) {
        Ecdsa.verify(pk, tc.hashAlg, tc.msg, tc.sig)
          .then(function(isValid) {
            if (
              (tc.result == "valid" && !isValid) ||
              (tc.result == "invalid" && isValid)
            ) {
              reject("Failed in test case " + tc.id);
            }
            resolve();
          })
          .catch(function(err) {
            // don't expect any exception in signature verification
            reject("Unexpected exception on test case " + tc.id + ": " + err);
          });
      })
      .catch(function(err) {
        reject("Failed to import key in test case " + tc.id + ": " + err);
      });
  });
  return promise;
};

/**
 * Parameters of a ECDSA signature verification test.
 * @param {!number} id Test case's id
 * @param {!JSONObject} keyData The key data in JWK format
 * @param {string} hashAlg The hash algorithm
 * @param {!ArrayBuffer} msg The message that was signed
 * @param {!ArrayBuffer} sig The signature to be verified
 * @param {string} result The expected result of the test case
 */
var EcdsaVerifyTestCase = function(id, keyData, hashAlg, msg, sig, result) {
  this.id = id;
  this.keyData = keyData;
  this.msg = msg;
  this.sig = sig;
  this.result = result;
  this.hashAlg = hashAlg;
};

/**
 * Tests ECDSA signature implementation with a number of test vectors.
 *
 * @return {!Promise}
 */
function testEcdsaVectors() {
  var tv = TestUtil.readJsonTestVectorsFromFile(ECDSA_VECTOR_FILE);
  var testCase = new goog.testing.TestCase();

  for (var i = 0; i < tv["testGroups"].length; i++) {
    var tg = tv["testGroups"][i];
    var keyData = tg["jwk"];
    var curveName = keyData["crv"];
    var hashAlg = tg["sha"];
    if (
      SUPPORTED["ecdsa-curve"].indexOf(curveName) == -1 ||
      SUPPORTED["hash"].indexOf(hashAlg) == -1
    ) {
      continue;
    }
    for (var j = 0; j < tg["tests"].length; j++) {
      var tc = tg["tests"][j];
      var tcId = tc["tcId"];
      var result = tc["result"];
      var msg = TestUtil.hexToArrayBuffer(tc["msg"]);
      var sig = TestUtil.hexToArrayBuffer(tc["sig"]);
      var test = new EcdsaVerifyTestCase(
        tcId,
        keyData,
        hashAlg,
        msg,
        sig,
        result
      );
      testCase.addNewTest("Test " + tcId, Ecdsa.testVerify, test);
    }
  }
  return testCase.runTestsReturningPromise().then(TestUtil.checkTestCaseResult);
}

/**
 * Extracts 'r' and 's' values from the given signature. The function assumes
 * that the given signature is the concatenation of 'r' and 's'.
 * @param {!ArrayBuffer} sig The signature
 *
 * @return {!Array<wycheproof.BigInteger>}
 */
Ecdsa.extractSig = function(sig) {
  var bytes = new Uint8Array(sig);
  var byteLen = bytes.length;
  var rBytes = bytes.subarray(0, byteLen / 2);
  var r = new BigInteger(rBytes);
  var sBytes = bytes.subarray(byteLen / 2, byteLen);
  var s = new BigInteger(sBytes);
  return [r, s];
};

/**
 * Extracts the random nonce that was used to calculate the given signature.
 * @param {!wycheproof.BigInteger} h The digest of the message
 * @param {!wycheproof.BigInteger} r The 'r' value of the signature
 * @param {!wycheproof.BigInteger} s The 's' value of the signature
 * @param {!wycheproof.BigInteger} d The private component of the ECDSA key
 * @param {!NistCurveSpec} curveSpec The specifications of the used curve
 *
 * @return {!wycheproof.BigInteger}
 */
Ecdsa.extractNonce = function(h, r, s, d, curveSpec) {
  var n = curveSpec.n;
  var k = d
    .multiply(r)
    .add(h)
    .multiply(n.modInverse(s))
    .mod(n);
  return k;
};

/**
 * Checks whether the given nonce was actually used during the signing process
 * of the given signature.
 * @param {!ArrayBuffer} msg The message that was signed
 * @param {!wycheproof.BigInteger} r The 'r' value of the signature
 * @param {!wycheproof.BigInteger} s The 's' value of the signature
 * @param {!wycheproof.BigInteger} d The private component of the ECDSA key
 * @param {!wycheproof.BigInteger} k The nonce that needs to be checked
 * @param {string} curveName The curve name
 */
Ecdsa.checkNonceCorrectness = function(msg, r, s, d, k, curveName) {
  var e2eCurveMap = {
    "P-256": "P_256",
    "P-384": "P_384",
    "P-521": "P_521"
  };
  var e2eCurveName = e2eCurveMap[curveName];
  var key = new e2e.ecc.Ecdsa(e2eCurveName, { privKey: d.toByteArray() });
  var msgBytes = new Uint8Array(msg);
  var calSig = key.signForTestingOnly(msgBytes, k);
  var calR = new BigInteger(calSig["r"]);
  var calS = new BigInteger(calSig["s"]);
  assertTrue(
    "Nonce calculation was incorrect",
    r.isEqual(calR) && s.isEqual(calS)
  );
};

/**
 * Tests whether there is bias in the nonce generation during
 * ECDSA signing process. The test case's parameters are passed
 * to 'this' variable of the function. The test is based on the fact that
 * if we throw a fair coin nTests times then the probability that
 * either heads or tails appears less than minCount is less than 2^{-32}.
 * Therefore the test below is not expected to fail unless the generation
 * of the one time keys is indeed biased.
 *
 * NOTE: This test only works correctly if the length of the output of the
 * hash function is equal to the bit length of the curve order.
 *
 * @return {!Promise}
 */
Ecdsa.testBias = function() {
  var tc = this;
  var countLsb = 0;
  var countMsb = 0;

  return Ecdsa.generateKey(tc.hashAlg, tc.curveName)
    .then(function(key) {
      return Ecdsa.exportKey(key.privateKey)
        .then(function(keyData) {
          var promises = [];

          for (var i = 0; i < tc.nTests; i++) {
            promises.push(
              Ecdsa.sign(key.privateKey, tc.msg, tc.hashAlg).then(function(
                sig
              ) {
                return HashUtil.digest(tc.hashAlg, tc.msg).then(function(
                  digest
                ) {
                  var curveSpec = EcUtil.getCurveSpec(tc.curveName);
                  var h = new BigInteger(new Uint8Array(digest));
                  // private key value
                  var d = BigInteger.fromHex(
                    TestUtil.base64UrlToHex(keyData["d"])
                  );
                  var r, s;
                  [r, s] = Ecdsa.extractSig(sig);
                  var k = Ecdsa.extractNonce(h, r, s, d, curveSpec);
                  // Uncomment this line to check correctness of
                  // nonce calculation
                  // Ecdsa.checkNonceCorrectness(tc.msg, r, s, d,
                  // k, tc.curveName);
                  var halfN = curveSpec.n.shiftRight(1);
                  if (k.isBitSet(0)) countLsb += 1;
                  if (k.compare(halfN) == 1) countMsb += 1;
                });
              })
            );
          }

          return Promise.all(promises).then(function() {
            if (countLsb < tc.minCount || countLsb > tc.nTests - tc.minCount) {
              reject(
                "Bias detected in the LSB of k" +
                  ", hash: " +
                  tc.hashAlg +
                  ", curve: " +
                  tc.curveName +
                  ", countLSB: " +
                  countLsb +
                  ", countMSB: " +
                  countMsb
              );
            }
            if (countMsb < tc.minCount || countMsb > tc.nTests - tc.minCount) {
              reject(
                "Bias detected in the MSB of k" +
                  ", hash: " +
                  tc.hashAlg +
                  ", curve: " +
                  tc.curveName +
                  ", countLSB: " +
                  countLsb +
                  ", countMSB: " +
                  countMsb
              );
            }
          });
        })
        .catch(function(err) {
          throw new Error("Failed to export private key: " + err);
        });
    })
    .catch(function(err) {
      throw new Error("Failed to generate key: " + err);
    });
};

/**
 * Parameters of a ECDSA bias test.
 * @param {string} hashAlg The hash algorithm
 * @param {string} curveName The curve name
 * @param {!ArrayBuffer} msg The message that was signed
 * @param {!number} nTests The number of key to be generated
 * @param {!number} minCount
 *   The expected number of times that a bit of the nonce should be 1 or 0
 */
var EcdsaBiasTestCase = function(hashAlg, curveName, msg, nTests, minCount) {
  this.msg = msg;
  this.hashAlg = hashAlg;
  this.curveName = curveName;
  this.nTests = nTests;
  this.minCount = minCount;
};

/**
 * Tests whether there is bias in the nonce generation during
 * ECDSA signing process. It considers three curves: secp256r1, secp384r1,
 * and secp521r1.
 *
 * @return {!Promise}
 */
function testEcdsaBiasAll() {
  var testCase = new goog.testing.TestCase();
  testCase.promiseTimeout = 120 * 1000;
  var msg = TestUtil.hexToArrayBuffer("48656c6c6f"); // msg = 'Hello'
  var nTests = 1024;
  var minCount = 410;
  var biasTest256 = new EcdsaBiasTestCase(
    "SHA-256",
    "P-256",
    msg,
    nTests,
    minCount
  );
  testCase.addNewTest("bias256", Ecdsa.testBias, biasTest256);
  var biasTest384 = new EcdsaBiasTestCase(
    "SHA-384",
    "P-384",
    msg,
    nTests,
    minCount
  );
  testCase.addNewTest("bias384", Ecdsa.testBias, biasTest384);
  var biasTest521 = new EcdsaBiasTestCase(
    "SHA-512",
    "P-521",
    msg,
    nTests,
    minCount
  );
  testCase.addNewTest("bias521", Ecdsa.testBias, biasTest521);
  return testCase
    .runTestsReturningPromise()
    .then(wycheproof.TestUtil.checkTestCaseResult);
}
